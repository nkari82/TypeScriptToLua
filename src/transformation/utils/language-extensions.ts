import * as ts from "typescript";
import { TransformationContext } from "../context";

export enum ExtensionKind {
    MultiFunction = "MultiFunction",
    RangeFunction = "RangeFunction",
    VarargConstant = "VarargConstant",
    AdditionOperatorType = "Addition",
    AdditionOperatorMethodType = "AdditionMethod",
    SubtractionOperatorType = "Subtraction",
    SubtractionOperatorMethodType = "SubtractionMethod",
    MultiplicationOperatorType = "Multiplication",
    MultiplicationOperatorMethodType = "MultiplicationMethod",
    DivisionOperatorType = "Division",
    DivisionOperatorMethodType = "DivisionMethod",
    ModuloOperatorType = "Modulo",
    ModuloOperatorMethodType = "ModuloMethod",
    PowerOperatorType = "Power",
    PowerOperatorMethodType = "PowerMethod",
    FloorDivisionOperatorType = "FloorDivision",
    FloorDivisionOperatorMethodType = "FloorDivisionMethod",
    BitwiseAndOperatorType = "BitwiseAnd",
    BitwiseAndOperatorMethodType = "BitwiseAndMethod",
    BitwiseOrOperatorType = "BitwiseOr",
    BitwiseOrOperatorMethodType = "BitwiseOrMethod",
    BitwiseExclusiveOrOperatorType = "BitwiseExclusiveOr",
    BitwiseExclusiveOrOperatorMethodType = "BitwiseExclusiveOrMethod",
    BitwiseLeftShiftOperatorType = "BitwiseLeftShift",
    BitwiseLeftShiftOperatorMethodType = "BitwiseLeftShiftMethod",
    BitwiseRightShiftOperatorType = "BitwiseRightShift",
    BitwiseRightShiftOperatorMethodType = "BitwiseRightShiftMethod",
    ConcatOperatorType = "Concat",
    ConcatOperatorMethodType = "ConcatMethod",
    LessThanOperatorType = "LessThan",
    LessThanOperatorMethodType = "LessThanMethod",
    GreaterThanOperatorType = "GreaterThan",
    GreaterThanOperatorMethodType = "GreaterThanMethod",
    NegationOperatorType = "Negation",
    NegationOperatorMethodType = "NegationMethod",
    BitwiseNotOperatorType = "BitwiseNot",
    BitwiseNotOperatorMethodType = "BitwiseNotMethod",
    LengthOperatorType = "Length",
    LengthOperatorMethodType = "LengthMethod",
    TableNewType = "TableNew",
    TableDeleteType = "TableDelete",
    TableDeleteMethodType = "TableDeleteMethod",
    TableGetType = "TableGet",
    TableGetMethodType = "TableGetMethod",
    TableHasType = "TableHas",
    TableHasMethodType = "TableHasMethod",
    TableSetType = "TableSet",
    TableSetMethodType = "TableSetMethod",
    TableAddKeyType = "TableAddKey",
    TableAddKeyMethodType = "TableAddKeyMethod",
}
const extensionValues: Set<string> = new Set(Object.values(ExtensionKind));

export function getExtensionKindForType(context: TransformationContext, type: ts.Type): ExtensionKind | undefined {
    const value = getPropertyValue(context, type, "__tstlExtension");
    if (value && extensionValues.has(value)) {
        return value as ExtensionKind;
    }
}

const excludedTypeFlags: ts.TypeFlags =
    ((1 << 18) - 1) | // All flags from Any...Never
    ts.TypeFlags.Index |
    ts.TypeFlags.NonPrimitive;

function getPropertyValue(context: TransformationContext, type: ts.Type, propertyName: string): string | undefined {
    if (type.flags & excludedTypeFlags) return;
    const property = type.getProperty(propertyName);
    if (!property) return undefined;
    const propertyType = context.checker.getTypeOfSymbolAtLocation(property, context.sourceFile);
    if (propertyType.isStringLiteral()) return propertyType.value;
}

export function getExtensionKindForNode(context: TransformationContext, node: ts.Node): ExtensionKind | undefined {
    const originalNode = ts.getOriginalNode(node);
    let type = context.checker.getTypeAtLocation(originalNode);
    if (ts.isOptionalChain(originalNode)) {
        type = context.checker.getNonNullableType(type);
    }
    return getExtensionKindForType(context, type);
}

export function getExtensionKindForSymbol(
    context: TransformationContext,
    symbol: ts.Symbol
): ExtensionKind | undefined {
    const type = context.checker.getTypeOfSymbolAtLocation(symbol, context.sourceFile);
    return getExtensionKindForType(context, type);
}

export enum IterableExtensionKind {
    Iterable = "Iterable",
    Pairs = "Pairs",
    PairsKey = "PairsKey",
}

export function getIterableExtensionTypeForType(
    context: TransformationContext,
    type: ts.Type
): IterableExtensionKind | undefined {
    const value = getPropertyValue(context, type, "__tstlIterable");
    if (value && value in IterableExtensionKind) {
        return value as IterableExtensionKind;
    }
}

export function getIterableExtensionKindForNode(
    context: TransformationContext,
    node: ts.Node
): IterableExtensionKind | undefined {
    const type = context.checker.getTypeAtLocation(node);
    return getIterableExtensionTypeForType(context, type);
}
