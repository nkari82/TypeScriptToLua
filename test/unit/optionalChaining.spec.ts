import { notAllowedOptionalAssignment } from "../../src/transformation/utils/diagnostics";
import * as util from "../util";
import { ScriptTarget } from "typescript";

test.each(["null", "undefined", '{ foo: "foo" }'])("optional chaining (%p)", value => {
    util.testFunction`
        const obj: {foo: string} | null | undefined = ${value};
        return obj?.foo;
    `
        .expectToMatchJsResult()
        .expectLuaToMatchSnapshot();
    // should use "and" expression
});

test("long optional chain", () => {
    util.testFunction`
        const a = { b: { c: { d: { e: { f: "hello!"}}}}};
        return a.b?.c?.d.e.f;
    `
        .expectToMatchJsResult()
        .expectLuaToMatchSnapshot();
    // should use "and" expression
});

test.each(["undefined", "{}", "{ foo: {} }", "{ foo: {bar: 'baz'}}"])("nested optional chaining (%p)", value => {
    util.testFunction`
        const obj: { foo?: { bar?: string } } | undefined = ${value};
        return obj?.foo?.bar;
    `.expectToMatchJsResult();
});

test.each(["undefined", "{}", "{ foo: {} }", "{ foo: {bar: 'baz'}}"])(
    "nested optional chaining combined with coalescing (%p)",
    value => {
        util.testFunction`
        const obj: { foo?: { bar?: string } } | undefined = ${value};
        return obj?.foo?.bar ?? "not found";
    `.expectToMatchJsResult();
    }
);

test.each(["[1, 2, 3, 4]", "undefined"])("optional array access (%p)", value => {
    util.testFunction`
        const arr: number[] | undefined = ${value};
        return arr?.[2];
    `.expectToMatchJsResult();
});

test.each(["[1, [2, [3, [4, 5]]]]", "[1, [2, [3, undefined]]] ", "[1, undefined]"])(
    "optional element access nested (%p)",
    value => {
        util.testFunction`
        const arr: [number, [number, [number, [number, number] | undefined]]] | [number, undefined] = ${value};
        return arr[1]?.[1][1]?.[0];
    `.expectToMatchJsResult();
    }
);

test.each(["{ }", "{ a: { } }", "{ a: { b: [{ c: 10 }] } }"])(
    "optional nested element access properties (%p)",
    value => {
        util.testFunction`
        const obj: {a?: {b?: Array<{c: number }> } } = ${value};
        return [obj["a"]?.["b"]?.[0]?.["c"] ?? "not found", obj["a"]?.["b"]?.[2]?.["c"] ?? "not found"];
    `.expectToMatchJsResult();
    }
);

test("optional element function calls", () => {
    util.testFunction`
        const obj: { value: string; foo?(this: void, v: number): number; bar?(this: void, v: number): number; } = {
            value: "foobar",
            foo: (v: number) => v + 10
        }
        const fooKey = "foo";
        const barKey = "bar";
        return obj[barKey]?.(5) ?? obj[fooKey]?.(15);
    `
        .expectToMatchJsResult()
        .expectLuaToMatchSnapshot();
    // should still use "and" statement, as functions have no self
});

test("unused expression", () => {
    util.testFunction`
        const obj = { foo: "bar" };
        obj?.foo;
    `
        .expectToHaveNoDiagnostics()
        .expectNoExecutionError()
        .expectLuaToMatchSnapshot();
    // should use if statement, as result is not used
});

test("unused call", () => {
    util.testFunction`
        let result
        const obj = {
            foo() {
                result = "bar"
            }
        };
        obj?.foo();
        return result;
    `
        .expectToMatchJsResult()
        .expectLuaToMatchSnapshot();
    // should use if statement, as result is not used
});

test.each(["undefined", "{ foo: v=>v }"])("with preceding statements on right side", value => {
    util.testFunction`
        let i = 0
        const obj: any = ${value};
        return {result: obj?.foo(i++), i};
    `
        .expectToMatchJsResult()
        .expectLuaToMatchSnapshot();
    // should use if statement, as there are preceding statements
});

// unused, with preceding statements on right side
test.each(["undefined", "{ foo(val) {return val} }"])(
    "unused result with preceding statements on right side",
    value => {
        util.testFunction`
        let i = 0
        const obj = ${value};
        obj?.foo(i++);
        return i
    `
            .expectToHaveNoDiagnostics()
            .expectLuaToMatchSnapshot();
        // should use if statement, as there are preceding statements
    }
);

test.each(["undefined", "{ foo(v) { return v} }"])("with preceding statements on right side modifying left", value => {
    util.testFunction`
        let i = 0
        let obj: any = ${value};
        function bar() {
            if(obj) obj.foo = undefined
            obj = undefined
            return 1
        }

        return {result: obj?.foo(bar(), i++), obj, i}
  `
        .expectToMatchJsResult()
        .expectLuaToMatchSnapshot();
    // should use if statement, as there are preceding statements
});

test("does not suppress error if left side is false", () => {
    const result = util.testFunction`
        const obj: any = false
        return obj?.foo
    `.getLuaExecutionResult();
    expect(result).toBeInstanceOf(util.ExecutionError);
});

describe("optional access method calls", () => {
    test("element access call", () => {
        util.testFunction`
        const obj: { value: string; foo?(prefix: string): string; bar?(prefix: string): string; } = {
            value: "foobar",
            foo(prefix: string) { return prefix + this.value; }
        }
        const fooKey = "foo";
        const barKey = "bar";
        return obj[barKey]?.("bar?") ?? obj[fooKey]?.("foo?");
    `.expectToMatchJsResult();
    });

    test("property access call", () => {
        util.testFunction`
        const obj: { value: string; foo?(prefix: string): string; bar?(prefix: string): string; } = {
            value: "foobar",
            foo(prefix: string) { return prefix + this.value; }
        }
        return obj.foo?.("bar?") ?? obj.bar?.("foo?");
    `.expectToMatchJsResult();
    });

    test("nested optional element access call", () => {
        util.testFunction`
        const obj: { value: string; foo?(prefix: string): string; bar?(prefix: string): string; } = {
            value: "foobar",
            foo(prefix: string) { return prefix + this.value; }
        }
        const fooKey = "foo";
        const barKey = "bar";
        return obj?.[barKey]?.("bar?") ?? obj?.[fooKey]?.("foo?");
    `.expectToMatchJsResult();
    });

    test("nested optional property access call", () => {
        util.testFunction`
        const obj: { value: string; foo?(prefix: string): string; bar?(prefix: string): string; } = {
            value: "foobar",
            foo(prefix: string) { return prefix + this.value; }
        }
        return obj?.foo?.("bar?") ?? obj?.bar?.("foo?");
    `.expectToMatchJsResult();
    });
});

test("no side effects", () => {
    util.testFunction`
        function getFoo(): { foo: number } | undefined {
            return { foo: 42 };
        }
        let barCalls = 0;
        function getBar(): { bar: number } | undefined {
            barCalls += 1;
            return undefined;
        }
        const result = getFoo()?.foo ?? getBar()?.bar;
        return { result, barCalls };
    `.expectToMatchJsResult();
});

// Test for https://github.com/TypeScriptToLua/TypeScriptToLua/issues/1044
test("does not crash when incorrectly used in assignment (#1044)", () => {
    const { diagnostics } = util.testFunction`
        foo?.bar = "foo";
    `.getLuaResult();

    expect(diagnostics.find(d => d.code === notAllowedOptionalAssignment.code)).toBeDefined();
});

describe("optional chaining function calls", () => {
    test.each(["() => 4", "undefined"])("stand-alone optional function (%p)", value => {
        util.testFunction`
            const f: (() => number) | undefined = ${value};
            return f?.();
        `.expectToMatchJsResult();
    });

    test("methods present", () => {
        util.testFunction`
            const objWithMethods = {
                foo() {
                    return 3;
                },
                bar(this: void) {
                    return 5;
                }
            };

            return [objWithMethods?.foo(), objWithMethods?.bar()];
        `.expectToMatchJsResult();
    });

    test("object with method can be undefined", () => {
        util.testFunction`
            const objWithMethods: { foo: () => number, bar: (this: void) => number } | undefined = undefined;
            return [objWithMethods?.foo() ?? "no foo", objWithMethods?.bar() ?? "no bar"];
        `.expectToMatchJsResult();
    });

    test("nested optional method call", () => {
        util.testFunction`
            type typeWithOptional = { a?: { b: { c: () => number } } };

            const objWithMethods: typeWithOptional = {};
            const objWithMethods2: typeWithOptional = { a: { b: { c: () => 4 } } };

            return {
                expectNil: objWithMethods.a?.b.c(),
                expectFour: objWithMethods2.a?.b.c()
            };
        `.expectToMatchJsResult();
    });

    test("methods are undefined", () => {
        util.testFunction`
            const objWithMethods: { foo?: () => number, bar?: (this: void) => number } = {};
            return [objWithMethods.foo?.() ?? "no foo", objWithMethods.bar?.() ?? "no bar"];
        `.expectToMatchJsResult();
    });

    test("optional method of optional method result", () => {
        util.testFunction`
            const obj: { a?: () => {b: {c?: () => number }}} = {};
            const obj2: { a?: () => {b: {c?: () => number }}} = { a: () => ({b: {}})};
            const obj3: { a?: () => {b: {c?: () => number }}} = { a: () => ({b: { c: () => 5 }})};

            return [obj.a?.().b.c?.() ?? "nil", obj2.a?.().b.c?.() ?? "nil", obj3.a?.().b.c?.() ?? "nil"];
        `.expectToMatchJsResult();
    });

    // https://github.com/TypeScriptToLua/TypeScriptToLua/issues/1085
    test("incorrect type, method is not a function", () => {
        util.testFunction`
            const obj: any = {};
            obj?.foo();
        `.expectToEqual(new util.ExecutionError("attempt to call a nil value (method 'foo')"));
    });

    describe("builtins", () => {
        test.each([
            ["undefined", undefined],
            ["{foo: 0}", true],
        ])("LuaTable: %p", (expr, value) => {
            util.testFunction`
            const table: LuaTable = ${expr} as any
            const bar = table?.has("foo")
            return bar
        `
                .withLanguageExtensions()
                .expectToEqual(value);
        });

        test.each(["undefined", "foo"])("Function call: %p", e => {
            util.testFunction`
            const result = []
            function foo(this: unknown, arg: unknown) {
                return [this, arg]
            }
            const bar = ${e} as typeof foo | undefined
            return bar?.call(1, 2)
        `.expectToMatchJsResult();
        });

        test.each([undefined, "[1, 2, 3, 4]"])("Array: %p", expr => {
            util.testFunction`
            const value: any[] | undefined = ${expr}
            return value?.map(x=>x+1)
        `.expectToMatchJsResult();
        });
    });

    test.each([true, false])("Default call context, strict %s", strict => {
        util.testFunction`
                function func(this: unknown, arg: unknown) {
                    return [this === globalThis ? "_G" : this === undefined ? "nil" : "neither", arg];
                }
                let i = 0
                const result = func?.(i++);
        `
            .setOptions({
                strict,
                target: ScriptTarget.ES5,
            })
            .expectToMatchJsResult();
    });
});

describe("Unsupported optional chains", () => {
    test("Language extensions", () => {
        util.testModule`
                new LuaTable().has?.(3)
            `
            .withLanguageExtensions()
            .expectDiagnosticsToMatchSnapshot();
    });

    test("Builtin prototype method", () => {
        util.testModule`
                [1,2,3,4].forEach?.(()=>{})
            `.expectDiagnosticsToMatchSnapshot();
    });

    test("Builtin global method", () => {
        util.testModule`
                Number?.("3")
            `.expectDiagnosticsToMatchSnapshot();
    });

    test("Builtin global property", () => {
        util.testModule`
                console?.log("3")
            `
            .setOptions({
                lib: ["lib.esnext.d.ts", "lib.dom.d.ts"],
            })
            .expectDiagnosticsToMatchSnapshot();
    });

    test("Compile members only", () => {
        util.testFunction`
                /** @compileMembersOnly */
                enum TestEnum {
                    A = 0,
                    B = 2,
                    C,
                    D = "D",
                }

                TestEnum?.B
            `.expectDiagnosticsToMatchSnapshot();
    });
});

describe("optional delete", () => {
    test("successful", () => {
        util.testFunction`
            const table = {
                bar: 3
            }
            return [delete table?.bar, table]
        `.expectToMatchJsResult();
    });

    test("unsuccessful", () => {
        util.testFunction`
            const table : {
                bar?: number
            } = {}
            return [delete table?.bar, table]
       `.expectToMatchJsResult();
    });

    test("delete on undefined", () => {
        util.testFunction`
            const table : {
                bar: number
            } | undefined = undefined
            return [delete table?.bar, table ?? "nil"]
        `.expectToMatchJsResult();
    });
});

describe("Non-null chain", () => {
    test("Single non-null chain", () => {
        util.testFunction`
            const foo = {a: { b: 3} }
            return foo?.a!.b
        `.expectToMatchJsResult();
    });

    test("Many non-null chains", () => {
        util.testFunction`
            const foo = {a: { b: 3} }
            return foo?.a!!!.b!!!
        `.expectToMatchJsResult();
    });
});
