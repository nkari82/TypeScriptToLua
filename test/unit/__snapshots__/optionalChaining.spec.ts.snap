// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unsupported optional chains Builtin global method: code 1`] = `
"local ____lualib = require(\\"lualib_bundle\\")
local __TS__Number = ____lualib.__TS__Number
local ____opt_0 = Number
if ____opt_0 ~= nil then
    __TS__Number(\\"3\\")
end"
`;

exports[`Unsupported optional chains Builtin global method: diagnostics 1`] = `"main.ts(2,17): error TSTL: Optional calls are not supported for builtin or language extension functions."`;

exports[`Unsupported optional chains Builtin global property: code 1`] = `
"local ____opt_0 = console
if ____opt_0 ~= nil then
    print(\\"3\\")
end"
`;

exports[`Unsupported optional chains Builtin global property: diagnostics 1`] = `"main.ts(2,17): error TSTL: Optional calls are not supported for builtin or language extension functions."`;

exports[`Unsupported optional chains Builtin prototype method: code 1`] = `
"local ____lualib = require(\\"lualib_bundle\\")
local __TS__ArrayForEach = ____lualib.__TS__ArrayForEach
local ____opt_0 = ({1, 2, 3, 4}).forEach
if ____opt_0 ~= nil then
    __TS__ArrayForEach(
        {1, 2, 3, 4},
        function()
        end
    )
end"
`;

exports[`Unsupported optional chains Builtin prototype method: diagnostics 1`] = `"main.ts(2,17): error TSTL: Optional calls are not supported for builtin or language extension functions."`;

exports[`Unsupported optional chains Compile members only: code 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    ---
    -- @compileMembersOnly
    local A = 0
    ---
    -- @compileMembersOnly
    local B = 2
    ---
    -- @compileMembersOnly
    local C = 3
    ---
    -- @compileMembersOnly
    local D = \\"D\\"
    local ____opt_0 = TestEnum
    if ____opt_0 ~= nil then
        local ____ = B
    end
end
return ____exports"
`;

exports[`Unsupported optional chains Compile members only: diagnostics 1`] = `"main.ts(10,17): error TSTL: Optional calls are not supported on enums marked with @compileMembersOnly."`;

exports[`Unsupported optional chains Language extensions: code 1`] = `
"local ____opt_0 = ({}).has
if ____opt_0 ~= nil then
end"
`;

exports[`Unsupported optional chains Language extensions: diagnostics 1`] = `
"main.ts(2,17): error TSTL: Optional calls are not supported for builtin or language extension functions.
main.ts(2,17): error TSTL: This language extension must be called as a method."
`;

exports[`long optional chain 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local a = {b = {c = {d = {e = {f = \\"hello!\\"}}}}}
    local ____opt_2 = a.b
    local ____opt_0 = ____opt_2 and ____opt_2.c
    return ____opt_0 and ____opt_0.d.e.f
end
return ____exports"
`;

exports[`optional chaining ("{ foo: \\"foo\\" }") 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local obj = {foo = \\"foo\\"}
    return obj and obj.foo
end
return ____exports"
`;

exports[`optional chaining ("null") 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local obj = nil
    return obj and obj.foo
end
return ____exports"
`;

exports[`optional chaining ("undefined") 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local obj = nil
    return obj and obj.foo
end
return ____exports"
`;

exports[`optional element function calls 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local obj = {
        value = \\"foobar\\",
        foo = function(v) return v + 10 end
    }
    local fooKey = \\"foo\\"
    local barKey = \\"bar\\"
    local ____opt_0 = obj[barKey]
    local ____temp_4 = ____opt_0 and ____opt_0(5)
    if ____temp_4 == nil then
        local ____opt_2 = obj[fooKey]
        ____temp_4 = ____opt_2 and ____opt_2(15)
    end
    return ____temp_4
end
return ____exports"
`;

exports[`unused call 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local result
    local obj = {foo = function(self)
        result = \\"bar\\"
    end}
    if obj ~= nil then
        obj:foo()
    end
    return result
end
return ____exports"
`;

exports[`unused expression 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local obj = {foo = \\"bar\\"}
    if obj ~= nil then
        local ____ = obj.foo
    end
end
return ____exports"
`;

exports[`unused result with preceding statements on right side 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local i = 0
    local obj = nil
    if obj ~= nil then
        local ____opt_0_foo_2 = obj.foo
        local ____i_1 = i
        i = ____i_1 + 1
        ____opt_0_foo_2(obj, ____i_1)
    end
    return i
end
return ____exports"
`;

exports[`unused result with preceding statements on right side 2`] = `
"local ____exports = {}
function ____exports.__main(self)
    local i = 0
    local obj = {foo = function(self, val)
        return val
    end}
    if obj ~= nil then
        local ____opt_0_foo_2 = obj.foo
        local ____i_1 = i
        i = ____i_1 + 1
        ____opt_0_foo_2(obj, ____i_1)
    end
    return i
end
return ____exports"
`;

exports[`with preceding statements on right side 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local i = 0
    local obj = nil
    local ____opt_result_4
    if obj ~= nil then
        local ____opt_0_foo_2 = obj.foo
        local ____i_1 = i
        i = ____i_1 + 1
        ____opt_result_4 = ____opt_0_foo_2(obj, ____i_1)
    end
    return {result = ____opt_result_4, i = i}
end
return ____exports"
`;

exports[`with preceding statements on right side 2`] = `
"local ____exports = {}
function ____exports.__main(self)
    local i = 0
    local obj = {foo = function(____, v) return v end}
    local ____opt_result_4
    if obj ~= nil then
        local ____opt_0_foo_2 = obj.foo
        local ____i_1 = i
        i = ____i_1 + 1
        ____opt_result_4 = ____opt_0_foo_2(obj, ____i_1)
    end
    return {result = ____opt_result_4, i = i}
end
return ____exports"
`;

exports[`with preceding statements on right side modifying left 1`] = `
"local ____exports = {}
function ____exports.__main(self)
    local i = 0
    local obj = nil
    local function bar(self)
        if obj then
            obj.foo = nil
        end
        obj = nil
        return 1
    end
    local ____opt_0 = obj
    if ____opt_0 ~= nil then
        local ____opt_0_foo_3 = ____opt_0.foo
        local ____bar_result_2 = bar(nil)
        local ____i_1 = i
        i = ____i_1 + 1
        ____opt_0 = ____opt_0_foo_3(____opt_0, ____bar_result_2, ____i_1)
    end
    return {result = ____opt_0, obj = obj, i = i}
end
return ____exports"
`;

exports[`with preceding statements on right side modifying left 2`] = `
"local ____exports = {}
function ____exports.__main(self)
    local i = 0
    local obj = {foo = function(self, v)
        return v
    end}
    local function bar(self)
        if obj then
            obj.foo = nil
        end
        obj = nil
        return 1
    end
    local ____opt_0 = obj
    if ____opt_0 ~= nil then
        local ____opt_0_foo_3 = ____opt_0.foo
        local ____bar_result_2 = bar(nil)
        local ____i_1 = i
        i = ____i_1 + 1
        ____opt_0 = ____opt_0_foo_3(____opt_0, ____bar_result_2, ____i_1)
    end
    return {result = ____opt_0, obj = obj, i = i}
end
return ____exports"
`;
